lib/streamlit/script_runner.py

```diff
@@ -295,9 +295,14 @@ def _run_script(self, rerun_data):
         # Run callbacks for widgets whose values have changed.
         if rerun_data.widget_states is not None:
             try:
-                self._widgets.call_callbacks(rerun_data.widget_states)
                 # Update the Widget object with the new widget_states.
+                # The old states, used to skip callbacks if values haven't changed,
+                # are preserved in widget_states.
+                # The state must be updated before the callbacks are run, so they can
+                # access the new states using beta_widget_state().
                 self._widgets.set_state(rerun_data.widget_states)
+
+                self._widgets.call_callbacks(rerun_data.widget_states)
```

Comment:
Curious...I wonder if these are related? Basically `set_state` should trigger `call_callbacks` That would require some investigation, so this is likely to be fine for now.

Reply:
Hmm, potentially. The api with my current changes is error-prone due to having some implicit ordering expectations. `set_state` triggering `call_callbacks` might be awkward if you need to set the state without triggering for some reason (and tying such things together feels ugly), but it would be safer/easier to use in the normal case, and feels kinda natural of a thing to do.